### ISSUE TITLE
Reuse material inside THREE.EdgesHelper

### ISSUE THREAD

**Issue by Wilt** [mrdoob/three.js#9123]
##### Description of the problem 

I am using `THREE.EdgesHelper` for virtually all objects inside my thee.js model (**r75**). Now I realized that for each edge-helper instance a new  `THREE.LineBasicMaterial` instance is created. Since I apply the same color to all my edges-helpers it would be more efficient to reuse the material instance once it is created. I can of course overwrite the material like this:

    var color = 0x000000;
    var material =  new THREE.LineBasicMaterial( { color: color } )
    var edgesHelper = new THREE.EdgesHelper( object, 0x000000, angle );

    edgesHelper.material = material;

##### Proposal/Question for optimizing this

Would it be an option to refactor the `THREE.EdgesHelper` so I can pass a material to the constructor so it can be reused in all my edges-helpers (I could make a PR for such a change)? Or would such possibility be undesired and should I continue solving my issue with the above code?

For example the second param could be a color or a material and if it is a material instance it will be used.

**Comment by WestLangley** [comment]
All the instances  of `EdgesHelper` are sharing the same shader.

**Comment by mrdoob** [comment]
Yeah, I'm not sure this is a optimization...

**Comment by Wilt** [comment]
@WestLangley could you please elaborate on this? How does that work? Apparently I am misunderstanding something fundamental about material usage...

I can also explain why the question popped up. I made a custom `SectionHelper` class (to show inside material of sliced geometries) based on this `EdgesHelper` class and i did like this:

````
THREE.SectionHelper = function ( object, color ) {

    THREE.Mesh.call( this, object.geometry, new THREE.MeshBasicMaterial( { color: color } ) );

    this.matrix = object.matrixWorld;
    this.matrixAutoUpdate = false;
};

THREE.SectionHelper.prototype = Object.create( THREE.Mesh.prototype );
THREE.SectionHelper.prototype.constructor = THREE.SectionHelper;
````

So in this case I create a new `THREE.MeshBasicMaterial` for SectionHelper instance. 
Then I realized that earlier in my project (at very beginning) I noticed a huge performance increase when I started reusing materials. That is why I rewrote the class to this:

````
THREE.SectionHelper = function ( object, hexOrMaterial ) {

    var material;
    if(hexOrMaterial instanceof THREE.MeshBasicMaterial) {
        material = hexOrMaterial;
    }else{
        var color = ( hexOrMaterial !== undefined ) ? hexOrMaterial : 0xffffff;
        material = new THREE.MeshBasicMaterial( { color: color } );
    }

    THREE.Mesh.call( this, object.geometry, material );

    this.matrix = object.matrixWorld;
    this.matrixAutoUpdate = false;

};

THREE.SectionHelper.prototype = Object.create( THREE.Mesh.prototype );
THREE.SectionHelper.prototype.constructor = THREE.SectionHelper;
````

How do these two cases (`EdgesHelper` with `LineBasicMaterial` and `SectionHelper` with `MeshBasicMaterial` relate to the issue of reusing material and improved performance. Does this performance issue only apply to `MeshBasicMaterial` and not to `LineBasicMaterial`. Or is this change in my `SectionHelper` redundant and also not an optimization.

In other words, when does reusing materials make sense and when not (it is would be great to enlighten me with some explanation as to understand this fundamental topic).

**Note:** Should I make a PR for this (simple) `SectionHelper` class. I saw people looking for such a thing earlier on StackOverflow it might be useful to others. Or are you not interested?

**Comment by mrdoob** [comment]
Have you done tests that show that your suggestion actually improves performance?

**Comment by mrdoob** [comment]
> @WestLangley could you please elaborate on this? How does that work?

Even if two objects have their own instance of `MeshBasicMaterial`, `WebGLRenderer` will only create one `program` internally.

**Comment by WestLangley** [comment]
@Wilt You will learn a lot from experimentation. Type `renderer.info` into the console to see -- among other things -- the number of programs that are created.

If you have many instances of `EdgesHelper`, there will be many material instances created on the CPU, so perhaps your mods can increase performance somewhat. Quantifying it would be a good idea.

**Comment by WestLangley** [comment]
@mrdoob 

>Even if two objects have their own instance of MeshBasicMaterial, WebGLRenderer will only create one program internally.

I am not surprised to see that this results in two programs because the `defines` are different:

	var material1 = new THREE.MeshBasicMaterial( { map: null } );
	var material2 = new THREE.MeshBasicMaterial( { map: texture } );

But I am surprised to see that this results in two programs, too:

	var material1 = new THREE.MeshBasicMaterial( { map: texture } );
	var material2 = new THREE.MeshBasicMaterial( { map: texture } );

**Comment by mrdoob** [comment]
I guess there is a bug 😕

**Comment by mrdoob** [comment]
Do you mind opening a new issue?

**Comment by Wilt** [comment]
@WestLangley This renderer.info is new for me... Learning new things every day :D 

@mrdoob Nope I don't mind opening a new issue at all, but I don't understand what bug you mean :P

**Comment by mrdoob** [comment]
> This is apparently a misconception.

Yep!
